File: src/main/java/net/casualuhc/uhcmod/UHCMod.java
package net.casualuhc.uhcmod;

import net.casualuhc.arcade.events.EventHandler;
import net.casualuhc.arcade.events.server.ServerLoadedEvent;
import net.casualuhc.arcade.recipes.RecipeHandler;
import net.casualuhc.arcade.scheduler.MinecraftTimeUnit;
import net.casualuhc.arcade.scheduler.Scheduler;
import net.casualuhc.uhcmod.features.GoldenHeadRecipe;
import net.casualuhc.uhcmod.managers.*;
import net.casualuhc.uhcmod.uhcs.ChristmasUHC;
import net.casualuhc.uhcmod.features.UHCAdvancements;
import net.casualuhc.uhcmod.uhcs.EasterUHC;
import net.casualuhc.uhcmod.utils.uhc.Config;
import net.casualuhc.uhcmod.utils.uhc.GenericEvents;
import net.casualuhc.uhcmod.utils.uhc.UHC;
import net.devtech.arrp.api.RRPCallback;
import net.devtech.arrp.api.RuntimeResourcePack;
import net.devtech.arrp.json.lang.JLang;
import net.fabricmc.api.ModInitializer;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

public class UHCMod implements ModInitializer {
    private static final Map<String, UHC> EVENTS = new HashMap<>();

    public static final Logger LOGGER = LoggerFactory.getLogger("UHC");

    public static MinecraftServer SERVER;

    static {
        GenericEvents.noop();

        EventHandler.register(ServerLoadedEvent.class, event -> {
            SERVER = event.getServer();
            Config.noop();

            // Fully load everything
            Scheduler.schedule(0, MinecraftTimeUnit.Ticks, () -> {
                UHCManager.noop();
                UHCAdvancements.noop();
                WorldBorderManager.noop();
                PlayerManager.noop();
                TeamManager.noop();
                RecipeManager.registerRecipes();

                Config.CURRENT_UHC.load();
            });
        });
    }

    public static void addUHCEvent(String string, UHC UHC) {
        EVENTS.put(string, UHC);
    }

    @Nullable
    public static UHC getUHC(String name) {
        return EVENTS.get(name);
    }

    @Override
    public void onInitialize() {
        addUHCEvent("christmas", new ChristmasUHC());
        addUHCEvent("easter", new EasterUHC());
    }
}

File: src/main/kotlin/net/casualuhc/uhcmod/UHCMod.kt
package net.casualuhc.uhcmod

import net.casualuhc.uhcmod.advancement.UHCAdvancements
import net.casualuhc.uhcmod.items.UHCItems
import net.casualuhc.uhcmod.managers.*
import net.casualuhc.uhcmod.minigame.Dimensions
import net.casualuhc.uhcmod.resources.UHCResourcePack
import net.casualuhc.uhcmod.resources.UHCResourcePackHost
import net.casualuhc.uhcmod.uhc.EasterUHC
import net.casualuhc.uhcmod.uhc.UHCEvents
import net.casualuhc.uhcmod.util.AntiCheat
import net.casualuhc.uhcmod.util.Config
import net.casualuhc.uhcmod.util.Texts
import net.fabricmc.api.ModInitializer
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.ModContainer
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class UHCMod: ModInitializer {
    companion object {
        const val ID = "uhc"

        val logger: Logger = LoggerFactory.getLogger("UHC")
        val uhc: ModContainer = FabricLoader.getInstance().getModContainer(this.ID).get()
    }

    override fun onInitialize() {
        UHCItems.noop()

        this.registerEvents()

        Texts.registerEvents()
        UHCManager.registerEvents()
        UHCAdvancements.registerEvents()
        PlayerManager.registerEvents()
        AntiCheat.registerEvents()
        CommandManager.registerEvents()
        TeamManager.registerEvents()
        DataManager.registerEvents()
        WorldBorderManager.registerEvents()
        UHCResourcePackHost.registerEvents()

        UHCResourcePack.initialise()

        Dimensions.noop()

        Config.registerEvents()
    }

    fun registerEvents() {
        UHCEvents.register("easter", ::EasterUHC)
    }

    // TODO:
    //   Fix Glow?
    //   Nerf World Borders
    //   EasterUHC - Add event UHCs
    //   Duels? - Implement duel command
    //   Testing - Run tests (implement unit tests?)
    //   Minesweeper AI - Make sure all maps are non-luck based
}